orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/CompleteExecutionHandler.kt

```diff
@@ -56,20 +56,24 @@ class CompleteExecutionHandler(
         log.info("Execution ${execution.id} already completed with ${execution.status} status")
       } else {
         message.determineFinalStatus(execution) { status ->
-          repository.updateStatus(execution.type, message.executionId, status)
-          publisher.publishEvent(
-            ExecutionComplete(this, message.executionType, message.executionId, status)
-          )
-          registry.counter(completedId.withTags(
-            "status", status.name,
-            "executionType", execution.type.name,
-            "application", execution.application
-          )).increment()
+          var runningStages: Boolean = false
           if (status != SUCCEEDED) {
             execution.topLevelStages.filter { it.status == RUNNING }.forEach {
+              runningStages = true
               queue.push(CancelStage(it))
             }
           }
+          if (!runningStages) {
```

Comment:
CompleteExecution gets queued at the end of every branch, and also requeued as needed, so this allows the other stages to cancel, complete their branch, and queue a CompleteExection.

Only once we don't have any stages to cancel do we update the status to its final value and publish the ExecutionComplete event

Reply:
Seems sane. Any danger of a race condition here? I don't think so.

Reply:
I also don't think so. So probably.